### ISSUE TITLE
adopt_atan2 does not work for this optimization problem

### ISSUE THREAD

**Issue by tesla3** [lucidrains/adam-atan2-pytorch#4]
I saw you are using AdoptAtan2 in your recent in RL codebase, so I assume you are successful there. But I encountered a failure case and want to report it so you are aware.

I tried both AdamAtan2 and AdoptAtan2 on boolean function in-context learning (repo: https://github.com/satwik77/incontext-bool). AdamAtan2 works very well.  But AdoptAtan2 doesn't , despite various different configurations.

Modification to `src/train.py`

```
filter out those that do not require grad
    param_dict = {pn: p for pn, p in param_dict.items() if p.requires_grad}
    # create optim groups. Any parameters that is 2D will be weight decayed, otherwise no.
    # i.e. all weight tensors in matmuls + embeddings decay, all biases and layernorms don't.
    decay_params = [p for n, p in param_dict.items() if p.dim() >= 2]
    nodecay_params = [p for n, p in param_dict.items() if p.dim() < 2]
    optim_groups = [
        {'params': decay_params, 'weight_decay': weight_decay},
        {'params': nodecay_params, 'weight_decay': 0.0}
    ]
    num_decay_params = sum(p.numel() for p in decay_params)
    num_nodecay_params = sum(p.numel() for p in nodecay_params)
    print(f"num decayed parameter tensors: {len(decay_params)}, with {num_decay_params:,} parameters")
    print(f"num non-decayed parameter tensors: {len(nodecay_params)}, with {num_nodecay_params:,} parameters")

    if optimizer == 'adopt_atan2':
        # adopt is insensitive to beta2
        #optimizer = AdoptAtan2(optim_groups, lr=args.learning_rate, betas=(betas[0], 0.999))
        optimizer = AdoptAtan2(model.parameters(), weight_decay=weight_decay, lr=args.learning_rate, betas=(betas[0], 0.999), cautious_factor=0.)
    elif optimizer == 'adam_atan2':
        # decoupled_wd does not work
        optimizer = AdamAtan2(optim_groups, lr=args.learning_rate, betas=betas)
    elif optimizer == 'adamw':
        # Create AdamW optimizer and use the fused version if it is available
        fused_available = 'fused' in inspect.signature(torch.optim.AdamW).parameters
        use_fused = fused_available and device_type == 'cuda'
        extra_args = dict(fused=True) if use_fused else dict()
        print(f"using fused AdamW: {use_fused}")
        optimizer = torch.optim.AdamW(optim_groups, lr=learning_rate, betas=betas, **extra_args)
    else:
        raise ValueError(f"Unknown optimizer {optimizer}")

    return optimizer

def train(model, args):
    optimizer = configure_optimizers(model, optimizer=args.optimizer, learning_rate=args.learning_rate, weight_decay=args.weight_decay)

```

reproduce by running (it does not converge.)
```
python -m src.train -project in-context-learning -name test -family san -model_name gpt2 -task conjunction -data boolean -train_steps 15000 -n_dims 28 -n_embd 256 -n_layer 12 -n_head 8 -batch_size 64 -learning_rate 0.00    1 -optimizer adopt_atan2 -weight_decay 0 -gpu 2
```
comparing to AdamAtan2
```
python -m src.train -project in-context-learning -name test -family san -model_name gpt2 -task conjunction -data boolean -train_steps 15000 -n_dims 28 -n_embd 256 -n_layer 12 -n_head 8 -batch_size 64 -learning_rate 0.0    001 -optimizer=adam_atan2 -gpu 3
```

and AdamW
```
python -m src.train -project in-context-learning -name test -family san -model_name gpt2 -task conjunction -data boolean -train_steps 15000 -n_dims 28 -n_embd 256 -n_layer 12 -n_head 8 -batch_size 64 -learning_rate 0.0    001 -optimizer=adamw -gpu 2
```

PS: I have tried following additional parameters and none converges
cautious_factor: 0.
weight_decay: 0.1, 0

**Comment by lucidrains** [comment]
hey, thank you so much for reporting this

I will take a look! have you tried cautious factor for adam atan2?

**Comment by tesla3** [comment]
I just tried cautious factor (0.), it converges. But I did not benchmark convergence speed.

**Comment by lucidrains** [comment]
🙏 let's stick with Adam atan2 then! made the changes in the other repo

**Comment by tesla3** [comment]
By the way, I also found, in this same problem, settting decoupled_wd in AdamAtan2 does not converge.

**Comment by lucidrains** [comment]
ohh, what value did you use for the decoupled wd? weight decay is pretty standard, it should work

**Comment by tesla3** [comment]
weight_decay=0.1.

**Comment by tesla3** [comment]
BTW: it is known decoupled weight decay helps widen stable learning rate range (https://fabian-sp.github.io/posts/2024/02/decoupling/) . I observed adam_atan2 widen the learning rate as well. They might have interfered each other.

**Comment by lucidrains** [comment]
@tesla3 ohhh yea, that's an insanely high weight decay 😆 try `1e-3`

**Comment by tesla3** [comment]
I remember that higher weight decay means "stronger L2 regularization".  In nanoGPT, wd is set to 0.1.

I will run one with lower WD and report back.

**Comment by lucidrains** [comment]
@tesla3 that seems high 🤔

ok, let me know, it should work

**Comment by tesla3** [comment]
You are right: bringing down wd to 0.001, decoupled_wd=True works. yeah!

**Comment by tesla3** [comment]
I tried both WD=0 and WD=0.01, neither converges.

```
# test 1
python -m src.train -project in-context-learning -name test -family san -model_name gpt2 -task conjunction -data boolean -train_steps 15000 -n_dims 28 -n_embd 256 -n_layer 12 -n_head 8 \
-batch_size 64 -learning_rate 0.001 -optimizer adopt_atan2 -weight_decay 0 -gpu 2

# test 2
python -m src.train -project in-context-learning -name test -family san -model_name gpt2 -task conjunction -data boolean -train_steps 15000 -n_dims 28 -n_embd 256 -n_layer 12 -n_head 8 \
-batch_size 64 -learning_rate 0.001 -optimizer adopt_atan2 -weight_decay 0.01 -gpu 3
```

**Comment by tesla3** [comment]
Change topic slightly, but still on optimizer. 

Have you tried Schedule-free optimizer (https://arxiv.org/abs/2405.15682).  In [this DL optimizer competition](https://openreview.net/forum?id=CtM5xjRSfm), it ranked top one out of box.  Plan to try it out. Let me know your thought.

**Comment by lucidrains** [comment]
@tesla3 interesting, will check out your task next week, thanks for running the experiments!

**Comment by lucidrains** [comment]
@tesla3 oh just noticed another issue, think you should be passing the `optim_groups` to constructor, so 

```python
optimizer = AdoptAtan2(optim_groups)
```

remove the cautious factor as well for better comparison

**Comment by tesla3** [comment]
Was thinking optim_groups vs model.parameters() shouldn't make a big difference here.  But,  it appears it does. 

```
        optimizer = AdoptAtan2(optim_groups, weight_decay=weight_decay, lr=learning_rate, betas=(betas[0], 0.999), cautious_factor=0.)
```
Now it converges with zero weight decay.
```
 -learning_rate 0.001 -optimizer adopt_atan2 -weight_decay 0 
```


But, it appears to converge slower than AdamAtan with similar configure (wd=0, decouple_wd=true) just eye balling..

**Comment by lucidrains** [comment]
@tesla3 so the 0. actually is cautious factor with maximum strength, you can set it to 1. or remove the line altogether to turn it off (confusing I know)
